components/shared/ListView/index.js

```diff
@@ -2,88 +2,410 @@ import React from "react";
-          </a>
-          {item.dataProvider &&
-            <span className={`${css.itemProvider}`}>
-              &nbsp; in {item.dataProvider}
-            </span>}
+class ListView extends React.Component {
+  state = {
+    readOnly: false,
+    listsInitialized: false,
+    listName: "",
+    listUUID: "",
+    selectedHash: {},
+    lists: [],
+    checkboxShown: false,
+    hasList: false,
+    listCreatedAt: 0,
+    showMessage: ""
+  };
+
+  componentDidMount() {
+    this.getLists();
+  }
+
+  getLists = async () => {
+    let readOnly = false;
+    this.setState({
+      listsInitialized: true
+    });
+    let lists = await getLocalForageLists();
+    // TODO: sometimes this doesnt arrive here because promises
+    lists.sort((a, b) => a.createdAt < b.createdAt);
+    if (this.props.defaultUUID) {
+      readOnly = true;
+      this.loadList(this.props.defaultUUID);
+    }
+    this.setState({
+      readOnly: readOnly,
+      lists: lists
+    });
+  };
+
+  onNameChange = value => {
+    this.createList(value);
+  };
+
+  createList = async listName => {
+    // add the new list to local storage
+    const uuid = createUUID();
+    const createdAt = Date.now();
+    let newLists = JSON.parse(JSON.stringify(this.state.lists));
+    newLists.push({
+      uuid: uuid,
+      name: listName,
+      count: 0,
+      createdAt: createdAt
+    });
+    newLists.sort((a, b) => a.createdAt < b.createdAt);
+    let savedList = {
+      name: listName,
+      selectedHash: {},
+      createdAt: createdAt,
+      updatedAt: createdAt
+    };
+    this.setState({
+      listName: listName,
+      listCreatedAt: createdAt,
+      listUUID: uuid,
+      selectedHash: {},
+      lists: newLists,
+      checkboxShown: true,
+      hasList: true
+    });
+    await setLocalForageItem(uuid, savedList);
+  };
+
+  loadList = async uuid => {
+    const value = await getLocalForageItem(uuid);
+    const listName = value.name;
+    const selectedHash = value.selectedHash;
+    const listCreatedAt = value.createdAt;
+
+    this.setState({
+      listName: listName,
+      listCreatedAt: listCreatedAt,
+      listUUID: uuid,
+      selectedHash: selectedHash,
+      checkboxShown: true,
+      hasList: true
+    });
+  };
+
+  listSelectChange = e => {
+    const listUUID = e.target.value;
+    if (listUUID === "") {
+      this.setState({
+        listName: "",
+        listUUID: "",
+        listCreatedAt: 0,
+        selectedHash: {},
+        checkboxShown: false,
+        hasList: false
+      });
+    } else {
+      this.loadList(listUUID);
+    }
+  };
+
+  onCheckItem = e => {
+    const element = e.target;
+    let id = element.getAttribute("data-id");
+    let selected = element.checked;
+    if (selected) {
+      this.addCell(id);
+    } else {
+      this.removeCell(id);
+    }
+  };
+
+  addCell = id => {
+    let hash = JSON.parse(JSON.stringify(this.state.selectedHash));
```

Comment:
Curious what this is... some sort of XSS prevention?

Reply:
i hear it is a faster/better way to make a deep copy of a hash in js https://stackoverflow.com/a/28876564/160933
i made an issue to refactor this into a lib https://github.com/dpla/dpla-frontend/issues/887

Reply:
It looks like in that SO post that they're suggesting you use the recursive approach... is that what you'd turn into the lib?

Why is there an lpad library but no deepcopy library? Oh javascript.

Reply:
true. i might make that function the lib when refactor